From fae55c41c75cf70eb96285b71a9405248a46c353 Mon Sep 17 00:00:00 2001
From: Jean Christophe Beyler <jean.christophe.beyler@intel.com>
Date: Tue, 30 Jul 2013 09:21:48 -0700
Subject: Dalvik: In dvmSetInitialBits correct bit index check and clear other bits

BZ: 127178

The bit number checked by dvmSetInitialBits is one greater than it
should be and the bits not set are left in an undefined state.
This patch corrects the off-by-one error and consequently must also check
whether the number of bits to set is 0 in order to avoid passing a bit
index of -1 to dvmSetBit.  It also adds functionality to explicitly clear
all bits that it does not set.

Category: device-enablement
Domain: AOSP-Dalvik-Runtime
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I3f5785458aec64d61cddbc5e96487183fb96f353
Orig-MCG-Change-Id: I67f85f7dd354152bc0d19027f5d4497485715fc8
Signed-off-by: Jean Christophe Beyler <jean.christophe.beyler@intel.com>
Signed-off-by: Eric Dittert <eric.dittert@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Reviewed-on: http://android.intel.com:8080/125081
Reviewed-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Tested-by: Chen, Dong-Yuan <dong-yuan.chen@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/BitVector.cpp |   21 ++++++++++++++++-----
 1 files changed, 16 insertions(+), 5 deletions(-)

diff --git a/vm/BitVector.cpp b/vm/BitVector.cpp
index aaa1a4d..b11392b 100644
--- a/vm/BitVector.cpp
+++ b/vm/BitVector.cpp
@@ -218,14 +218,21 @@ void dvmClearAllBits(BitVector* pBits)
 }
 
 /*
- * Mark specified number of bits as "set". Cannot set all bits like ClearAll
- * since there might be unused bits - setting those to one will confuse the
- * iterator.
+ * Mark specified number of bits as "set" and clear the rest.  Don't set all
+ * bits like ClearAll since there might be unused bits - setting those to one
+ * would confuse the iterator.
  */
 bool dvmSetInitialBits(BitVector* pBits, unsigned int numBits)
 {
-    //First step is to see if we can set those bits
-    if (dvmSetBit (pBits, numBits, false) == false)
+    //If numBits == 0, clear all the bits and report success
+    if (numBits == 0)
+    {
+        dvmClearAllBits (pBits);
+        return true;
+    }
+
+    //Otherwise, first step is to see if we can set those bits
+    if (dvmSetBit (pBits, numBits - 1, false) == false)
     {
         //Then don't do it
         return false;
@@ -238,6 +245,10 @@ bool dvmSetInitialBits(BitVector* pBits, unsigned int numBits)
     unsigned int remNumBits = numBits & 0x1f;
     if (remNumBits) {
         pBits->storage[idx] = (1 << remNumBits) - 1;
+        idx++;
+    }
+    for (; idx < pBits->storageSize; idx++) {
+        pBits->storage[idx] = 0;
     }
 
     //Report success
-- 
1.7.4.1

